labs/mokka-bft.md

```diff
@@ -0,0 +1,32 @@
+---
+layout: default
+title: Mokka-BFT
+parent: Labs
+---
+# Lab Name
+
+Mokka-BFT
+
+# Short Description
+
+Mokka-BFT Hyperledger lab is a repository for development of a production version of a BFT library based on the Mokka-BFT protocol,
+described in the paper 'Mokka: BFT consensus' (https://arxiv.org/ftp/arxiv/papers/1901/1901.08435.pdf).
```

Comment:
The paper reads: A node remains in a follower state as long as it receives a valid ACK packet from the leader. How does it tolerate a Byzantine leader that keeps sending ACK packets to all follower nodes, but doesn't fully follow the rest of the protocol?

Reply:
Hi there! There is a "safety section" in Mokka's whitepaper (will duplicate part of the section here): 

The leader uses the ACK packet to maintain its state. At least two malicious actions are possible here: replay attack and fake leader. 

The replay attack states that the same packet can be reused several times. This is crucial as the leader holds the network from the next voting (as he is still alive). So, a replay attack can be used to freeze the network. To overcome this issue, Mokka uses proof expiration (i.e., nonce): even if the malicious node tries replay attack, it will work until proof expires. In a production system, it is recommended to set expiration as low as possible, but keep in mind that each voting and leader switch will cause a temporary network hold (without a leader, it is impossible to commit new changes).

Fake leader attack states that fake node can send ACK packets and represent itself as a leader. To solve this issue, Mokka uses special proof (proof of voting), a multisig signature that is unique per term, timestamp, and candidate’s public key – so reuse in the long term is impossible. Acting without proof is impossible as well.

Reply:
Also, there is an option to bypass the await of proof expiration (in case the proof has been stolen). The solution is represented in "extensions to the original algorithm". If in short - we can introduce digital signatures for each ACK packet, however, this will reduce the sync speed, since validation of ECC signature takes time (in Node.js implementation around 4-5ms)

Reply:
I think there is some misunderstanding. I meant the behavior of a Byzantine leader when it tries to follow the protocol as little as possible, so that it just remains the leader, but prevents normal progress of the network.

By the way, how many Byzantine node can this protocol tolerate?

Reply:
Oh got the question. The only way to become the Byzantine leader here - is to steal the private key. In this way, you actually replace the original node with a fake one. As the algorithm is log-less, then stealing private key can lead to network freeze only, however, we can introduce the leader timeout (smth like one node can't be a leader 3 times in a row). 
Also, if we talk about logs, then we can extend the algorithm to validate the changes before apply on the state machine level (for instance validate digital signature against the transaction).

So, if we assume, that our node can be cracked from the inside (and private key can be stolen), then in any way we have to validate the logs on the follower's side.

Reply:
> The only way to become the Byzantine leader here - is to steal the private key.

I am still not quite sure that we got exactly to terms here.

The proposal name includes **BFT** which, I suppose, stands for [Byzantine Fault](https://en.wikipedia.org/wiki/Byzantine_fault) Tolerant. Byzantine fault is commonly understood synonymous to _arbitrary_ behavior. Unfortunately, I couldn't find a conventional description of neither the system model nor the guaranteed safety and liveness properties in the paper. Therefore, I assume the common system model and properties as in [PBFT'99 paper](http://pmg.csail.mit.edu/papers/osdi99.pdf). 

Achieving distributed consensus with Byzantine fault tolerance in the common model is well known to be a notoriously hard problem comparing to crash fault tolerance, even when employing modern cryptography and trusted hardware. Although it is relatively easy to achieve safety properties (nothing undesired happens), the biggest challenge is achieving liveness (the desired things eventually happen). 

I believe, to avoid confusion, any protocol that claims Byzantine fault tolerance has to guarantee both safety and liveness properties assuming some fraction of nodes can behave arbitrarily. This is why I was concerned how this protocol can tolerate arbitrary behavior of some nodes. For instance, a malicious leader may attempt to retain its leadership, but not fully follow the rest of the protocol, e.g. by intentionally not serving certain requests.

Reply:
I share @sergefdrv 's confusion here. I'm not a consensus expert but from a quick look at the paper it really isn't clear to me that this is BFT. Has this paper gone through any peer review? It seems to be lacking some of the information typically found in this kind of papers, like the number of faulty nodes that it can tolerate. There is also no information on performance, how does it compare to RAFT which it is said to be based on?

Note that this doesn't mean there is no value in the proposal, but it seems like a renaming is in order. Just calling it  something like "Mokka consensus protocol" would avoid the confusion.